File: inkbridge/src/screenshot.js
import { chromium } from 'playwright';
import sharp from 'sharp';

let browserPromise;
let shutdownHooksRegistered = false;

function registerShutdownHooks() {
  if (shutdownHooksRegistered) {
    return;
  }

  shutdownHooksRegistered = true;

  const shutdown = async () => {
    if (!browserPromise) {
      return;
    }

    try {
      const browser = await browserPromise;
      await browser.close();
    } catch (error) {
      console.error(`[browser] Failed to close Chromium during shutdown: ${error.message}`);
    } finally {
      browserPromise = undefined;
    }
  };

  process.once('beforeExit', shutdown);
  process.once('SIGINT', shutdown);
  process.once('SIGTERM', shutdown);
}

async function getBrowser() {
  if (!browserPromise) {
    browserPromise = chromium
      .launch({
        executablePath: '/usr/bin/chromium',
        args: ['--no-sandbox', '--disable-setuid-sandbox'],
      })
      .catch((error) => {
        browserPromise = undefined;
        throw error;
      });

    registerShutdownHooks();
  }

  return browserPromise;
}

async function captureScreenshot(slug, url, width, height, renderDelay, zoom, homeAssistantConfig) {
  const browser = await getBrowser();

  const context = await browser.newContext({
    viewport: { width, height },
  });

  try {
    const pageHostname = new URL(url).hostname;
    const homeAssistantHostname = new URL(homeAssistantConfig.url).hostname;
    const isHomeAssistant = pageHostname === homeAssistantHostname;

    if (isHomeAssistant) {
      console.log(`[${slug}] Home Assistant hostnames match. Injecting authentication tokens into localStorage.`);

      const hassTokens = {
        hassUrl: homeAssistantConfig.url,
        access_token: homeAssistantConfig.token,
        token_type: 'Bearer',
      };

      await context.addInitScript(
        (serializedTokens, selectedLanguage) => {
          window.localStorage.setItem('hassTokens', serializedTokens);
          window.localStorage.setItem('selectedLanguage', selectedLanguage);
        },
        JSON.stringify(hassTokens),
        JSON.stringify(homeAssistantConfig.language)
      );
    }

    const page = await context.newPage();
    await page.goto(url, { waitUntil: 'load' });

    await page.evaluate((pageZoom) => {
      document.body.style.zoom = String(pageZoom);
    }, zoom);

    if (isHomeAssistant) {
      await page.waitForSelector('home-assistant, ha-panel-lovelace', {
        state: 'visible',
        timeout: 2500,
      });
    }

    if (renderDelay > 0) {
      await page.waitForTimeout(renderDelay);
    }

    const buffer = await page.screenshot({ type: 'png' });
    const { data: rawBuffer, info } = await sharp(buffer).ensureAlpha().raw().toBuffer({ resolveWithObject: true });

    return { rawBuffer, info };
  } finally {
    await context.close();
  }
}

export { captureScreenshot };

File: inkbridge/src/screenshot.py
import json
from io import BytesIO
from urllib.parse import urlparse

from PIL import Image
from playwright.sync_api import Error as PlaywrightError, sync_playwright

from config import HomeAssistantConfig


def capture_screenshot(
    slug: str,
    url: str,
    width: int,
    height: int,
    render_delay: int,
    zoom: float,
    home_assistant_config: HomeAssistantConfig,
) -> Image.Image:
    with sync_playwright() as playwright:
        page_hostname = urlparse(url).hostname
        home_assistant_hostname = urlparse(home_assistant_config.url).hostname
        is_home_assistant = page_hostname == home_assistant_hostname
        storage_state = None

        if is_home_assistant:
            hass_tokens = {
                "hassUrl": home_assistant_config.url,
                "access_token": home_assistant_config.token,
                "token_type": "Bearer",
            }
            hass_origin = urlparse(home_assistant_config.url)
            storage_state = {
                "cookies": [],
                "origins": [
                    {
                        "origin": f"{hass_origin.scheme}://{hass_origin.netloc}",
                        "localStorage": [
                            {"name": "hassTokens", "value": json.dumps(hass_tokens)},
                            {"name": "selectedLanguage", "value": json.dumps(home_assistant_config.language)},
                        ],
                    }
                ],
            }

        browser = playwright.chromium.launch(
            executable_path="/usr/bin/chromium",
            args=["--no-sandbox", "--disable-setuid-sandbox"],
        )
        context = browser.new_context(viewport={"width": width, "height": height}, storage_state=storage_state)

        try:
            page = context.new_page()
            page.goto(url, wait_until="domcontentloaded")
            page.wait_for_load_state("load")

            page.evaluate("pageZoom => { document.body.style.zoom = String(pageZoom); }", zoom)

            if render_delay > 0:
                page.wait_for_timeout(render_delay)

            png_bytes = page.screenshot(type="png")
            return Image.open(BytesIO(png_bytes)).convert("RGBA")
        finally:
            context.close()
            browser.close()
